psycopg/psycopg/cursor.py

```diff
@@ -223,34 +229,16 @@ def _executemany_gen(
             else:
                 pgq.dump(params)
 
-            results = yield from self._maybe_prepare_gen(pgq, prepare=True)
-
-            if self._conn._pipeline:
-                yield from self._conn._pipeline._communicate_gen()
-            else:
-                assert results is not None
-                self._check_results(results)
-                if returning and results[0].status == ExecStatus.TUPLES_OK:
-                    self._results.extend(results)
-
-                for res in results:
-                    nrows += res.command_tuples or 0
-
-        if not self._conn._pipeline:
-            if self._results:
-                self._set_current_result(0)
-
-            # Override rowcount for the first result. Calls to nextset() will
-            # change it to the value of that result only, but we hope nobody
-            # will notice.
-            # You haven't read this comment.
-            self._rowcount = nrows
+            yield from self._maybe_prepare_gen(pgq, prepare=True)
+            yield from pipeline._communicate_gen()
 
         self._last_query = query
 
         for cmd in self._conn._prepared.get_maintenance_commands():
             yield from self._conn._exec_command(cmd)
 
+        yield from pipeline._flush_gen()
```

Comment:
Is this needed when not "returning"?

```python
        if self._execmany_returning:
            yield from pipeline._flush_gen()
```
maybe?

Reply:
Maybe it's not. However maybe it would be worth changing the default of returning to False, it would be a better win.

`returning=True` is useful only for select-like statements, in the sense that, even if `returning=True` but the statement doesn't return rows, the results are discarded from the cursor state (after getting their rowcount).

If there is an happy path which consists in not sending the flush, then it would be nice to hit it by default. Normally executemany is mostly used for updates and not to read the result, so it's reasonable for users to opt-it-in:

```python
cur.executemany("insert into tbl ... returning id". data, returning=True)
```

Also nice would have been to know upfront if the statement will return or not. We might do it by reading the result after sending the first record: if we see that it's a no-tuple result we can go not-returning, but this would stall the pipeline.

Someone had expressed [nervousness](https://github.com/psycopg/psycopg/issues/164#issuecomment-1044272063) in the default of returning being True. We may as well go conservative moving the default to False (it's not been released yet) and at the same time squeeze even more performance.

Reply:
Independently of `returning` being `True` or not, we still have to retrieve results in order to check if the operations succeeded. But, yes, we could avoid storing them.

Whatever `returning`'s value is, a flush will occur either when calling `cur.fetch*()` or/and when exiting the pipeline.

So by symmetry with the non-executemany case, this flush (even conditional) here is not needed, I would say. (Or we should emit one as well in `_execute_gen()`.)

Reply:
Thank you very much for the insight. I will test a bit with it :)

Reply:
Yes, I have tried dropping it and the test suite passes as before. I think it was part of a failed attempt to fix rowcount.

The trace doesn't change from what posted above, except that now the 
```
F	4	Flush
```

is missing. This probably helps to keep the pipeline flowing if there are more than just one executemany operation in it.